Testing message implementation

boekhorstb1 opened the pull request:
Adding public function checkingAsciiErrata() to MessageImplementation.php
based on this erratum: https://github.com/php-fig/fig-standards/pull/1274/files

tests are added to test/ServerRequestTest.php

boekhorstb1 committed:
setting bostrap to folder test and adding colors

boekhorstb1 committed:
first trial on testing something

boekhorstb1 committed:
updating my understanding of the problem

boekhorstb1 committed:
testing for psr4 errata

boekhorstb1 committed:
testing for the character code int the strings

boekhorstb1 committed:
backing up current code

boekhorstb1 committed:
use the following preg_matches to hit all ascii characters bellow 32

boekhorstb1 committed:
sw

boekhorstb1 committed:
creating unittest for serverrequest

boekhorstb1 committed:
updating tests and writign code for Messageimplementation trait

boekhorstb1 committed:
created private function checkingAsciiErrata for dry

boekhorstb1 committed:
updating MessageImplementation and ServerRequestTest

midahp commented on src/MessageImplementation.php:
```diff
@@ -220,20 +222,56 @@ public function withHeader($name, $value)
         return $ret;
     }
 
+    /** Adding checks following this errata: 1ttps://github.com/php-fig/fig-standards/pull/1274/files
```
link misses a `h` at the beginning

midahp commented on phpunit.xml.dist:
```diff
@@ -1,7 +1,8 @@
 <?xml version="1.0" encoding="UTF-8"?>
```
If you want to update this file maybe it would be best to use the one in horde/core

boekhorstb1 replied:
yes, ok will look that lateron... for some reason phpunit is not wokring when I do not make that change

midahp commented on src/MessageImplementation.php:
```diff
@@ -220,20 +222,56 @@ public function withHeader($name, $value)
+     * @param string $namevalue Header name or value(s).
+     * @throws \InvalidArgumentException When the body is not valid.
+     */
+    public function checkingAsciiErrata($namevalue)
```
That name is not ideal.
I would use something like `checkHeaderForInvalidAsciiChars`.
A bit long but more fitting

midahp replied:
Parameter should be `string $nameOrValue`
You could also have two arguemnts `string $name, string $value`. Then the error message could include where the error was.
But you can decide.

midahp commented on src/MessageImplementation.php:
```diff
@@ -220,20 +222,56 @@ public function withHeader($name, $value)
+        if (preg_match_all('/[\x00-\x20]/', $namevalue, $output)) {
+            # REJECT THE REQUEST
+            $erronousAsciiCharacters = '';
+            $outputarray = [];
```
`$outputArray` please

midahp commented on test/ServerRequestTest.php:
```diff
@@ -44,6 +46,36 @@ public function testCreateRelativeUrlRequest()
+    public function testHeaderThrowsExceptionWhenAsciiCharactersTill32InValue()
+    {   // This request should be refused due to invalid ascii characters in $headerValue
+        $this->expectException(InvalidArgumentException::class);
+        $this->expectDeprecationMessageMatches('/1, 5, 10, 32$/');
```
I am not sure what this does. Can you explain?

boekhorstb1 replied:
The function `checkHeaderForInvalidAsciiChars($namevalue)` throws e.g. the following error:

````
Failed asserting that exception message 'Invalid Characters found in header name. 
Please make sure to add headers correctly. 
Following invalid ASCII character-codes are found: 01, 05, 0a, 20
````
The test above checks if the correct error is thrown by checking the class and by checking if the correct hexadecimals have been found by the regex.

boekhorstb1 replied:
yes :) changed to hex

midahp commented on test/ServerRequestTest.php:
```diff
@@ -44,6 +46,36 @@ public function testCreateRelativeUrlRequest()
+        $this->expectDeprecationMessageMatches('/1, 5, 10, 32$/');
+        $headerName = 'TestHeader';
+        $headerValue = 'Trestvalue';
+        $headerValue .=  chr(1);
+        $headerValue .=  chr(5);
+        $headerValue .=  chr(10);
+        $headerValue .=  chr(32);
```
can you use the hex values here too or does that not work here?
It would make it more clear that you are testing the same values as above

boekhorstb1 replied:
yes :) changed to hex

midahp reviewed (commented):
Looks good. Just some small things.

boekhorstb1 committed:
updating values to hexvalues everywhere

boekhorstb1 committed:
sw

boekhorstb1 committed:
sc

boekhorstb1 committed:
returning phpunit.xml .dist o previous

boekhorstb1 committed:
sw

boekhorstb1 committed:
testing headers with response class

boekhorstb1 commented:
Question:
- I think (not sure) that there is a difference between explicitly setting 0x0a or \n... because: when I set 'blabla\nblabla' it is not matched by 0x0a but only by looking for \n ....
So: Why is that? Maybe that is why in the [link to the errata ](https://github.com/php-fig/fig-standards/pull/1274/files)explicitly \n and \r are mentioned. 

Im checking this link: https://www.php.net/manual/en/language.types.string.php
Noted that it might depends on how strings are defined. 

> If the string is enclosed in double-quotes ("), PHP will interpret the following escape sequences for special characters: 
> \n | linefeed (LF or 0x0A (10) in ASCII)
> \r | carriage return (CR or 0x0D (13) in ASCII)
>
> To specify a literal single quote, escape it with a backslash (\). To specify a literal backslash, double it (\\). All other instances of  backslash will be treated as a literal backslash: this means that the other escape sequences you might be used to, such as \r or \n, will be output literally as specified rather than having any special meaning.

midahp commented on src/MessageImplementation.php:
```diff
@@ -220,20 +222,56 @@ public function withHeader($name, $value)
+     * @param string $namevalue Header name or value(s).
+     * @throws \InvalidArgumentException When the body is not valid.
+     */
+    public function checkHeaderForInvalidAsciiChars($nameOrValue)
```
In https://github.com/php-fig/fig-standards/pull/1274/files header names and values have different character checks.
This needs to be adjusted. For that you should create two methods for checking.

boekhorstb1 committed:
setting less checks for header value

boekhorstb1 committed:
fix for checking each header value

boekhorstb1 committed:
changing tests for header value

boekhorstb1 committed:
updating ResponseTest.php

boekhorstb1 committed:
updating tests and removing filtering for unique characters

boekhorstb1 commented:
Not sure what to do about this part:

> If compatibility with older systems is desired then the sequence \r\n (0x0D0A) within a header value MAY be accepted if and only if it is immediately followed by either SPACE (0x20) or \t (0x09). The full sequence SHOULD then internally be normalized to a single SPACE (0x20).

midahp commented on src/MessageImplementation.php:
```diff
@@ -220,20 +222,69 @@ public function withHeader($name, $value)
+    {
+        // Checking headers name
+        if (isset($name) && preg_match_all('/[\x00-\x20]/', $name, $output)) {
+            # REJECT THE REQUEST
+            $erronousAsciiCharacters = '';
+            $outputArray = [];
+            foreach ($output[0] as $value) {
+                $outputArray[] = bin2hex($value);
+            }
+            $erronousAsciiCharacters = implode(', ', $outputArray);
+            throw new InvalidArgumentException('Invalid Characters found in header name. Please make sure to add headers correctly. Following invalid ASCII character-codes are found: '.$erronousAsciiCharacters);
```
This block is almost the same twice. Please refactor it in some way.
Try to repeat as little as possible.

midahp commented on src/MessageImplementation.php:
```diff
@@ -220,20 +222,69 @@ public function withHeader($name, $value)
+     * - `\n` (0x0A)
+     * - Any character less than or equal to 0x20.
+     *
+     * @param string $namevalue Header name or value(s).
```
update docstring. This is still for the old method

midahp commented on src/MessageImplementation.php:
```diff
@@ -220,20 +222,69 @@ public function withHeader($name, $value)
+     * @param string $namevalue Header name or value(s).
+     * @throws \InvalidArgumentException When the body is not valid.
+     */
+    public function checkHeaderForInvalidAsciiChars($name, $value)
```
please add types for the params `$name` and `$value`

midahp replied:
Types still missing here

midahp commented on src/MessageImplementation.php:
```diff
@@ -220,20 +222,69 @@ public function withHeader($name, $value)
+     * - Any character less than or equal to 0x20.
+     *
+     * @param string $namevalue Header name or value(s).
+     * @throws \InvalidArgumentException When the body is not valid.
```
Backslash not needed

midahp reviewed (commented):
Can you also add a test that checks that a header value with certain characters between 0x00 and 0x20 is valid?

midahp commented:
> Not sure what to do about this part:
> 
> > If compatibility with older systems is desired then the sequence \r\n (0x0D0A) within a header value MAY be accepted if and only if it is immediately followed by either SPACE (0x20) or \t (0x09). The full sequence SHOULD then internally be normalized to a single SPACE (0x20).

I think you can ignore that for now. But make sure that horde still works with the changes you made.

boekhorstb1 committed:
Refactoring code withanonymous function to makeit dry

boekhorstb1 committed:
updating docstring

boekhorstb1 committed:
sw

boekhorstb1 committed:
Adding test for not filtered characters

boekhorstb1 commented:
Ok I added another test to check the non-filtered characters are allowed ;)
Changed docstring and refactored code with help of anonymous nested function. 

I clicked around in the dev setup, I can login to master and configure things and play around in the simportal. So I think headers are working.

midahp commented on src/MessageImplementation.php:
```diff
@@ -220,20 +222,77 @@ public function withHeader($name, $value)
+    public function checkHeaderForInvalidAsciiChars($name, $value)
+    {
+        $checking = function ($output) {
+            $erronousAsciiCharacters = '';
```
This doesn't do anything

midahp commented on src/MessageImplementation.php:
```diff
@@ -220,20 +222,77 @@ public function withHeader($name, $value)
+        };
+
+        // Checking headers name
+        if (isset($name) && preg_match_all('/[\x00-\x20]/', $name, $output)) {
```
name should always be set. Otherwise an php error is thrown. At least if you define the param type, as I mentioned above.

midahp commented on src/MessageImplementation.php:
```diff
@@ -220,20 +222,77 @@ public function withHeader($name, $value)
+        }
+
+        // Checking headers value
+        if (isset($value)) {
```
Not needed (see above)

midahp commented on src/MessageImplementation.php:
```diff
@@ -220,20 +222,77 @@ public function withHeader($name, $value)
+                if (preg_match_all('/[\x00\x0D\x0A]/', $headervalue, $output)) {
+                    // reject request
+                    $erronousAsciiCharacters = $checking($output);
+                    throw new InvalidArgumentException('Invalid Characters found in header value. Please make sure to add headers correctly. Following invalid ASCII character-codes are found: '.$erronousAsciiCharacters);
```
This does not need to be so long.
`Found the following invalid ASCII character-codes in header name: ` should be enough.

midahp commented on src/MessageImplementation.php:
```diff
@@ -220,20 +222,77 @@ public function withHeader($name, $value)
+     */
+    public function checkHeaderForInvalidAsciiChars($name, $value)
+    {
+        $checking = function ($output) {
```
This function should have a different name. It is just a temporary function but it should still be obvious what it does.
Also, it might be useful to include the regex matching in this function

midahp reviewed (commented):
Overall good.
There are still some code quality issues but nothing major.

boekhorstb1 committed:
setting types, shortening strings and removing unnecessary code

boekhorstb1 committed:
changing function name
